### PR TITLE
DATAUP-292 import cell skeleton

### PR DESCRIPTION
**Pull request by briehl**
# Description of PR purpose/changes
This introduces a skeletal bulk import cell.
Most of what's here is (sort of) boilerplate. There's two parts to it.

First is the new bulk import cell extension. This is under nbextensions/bulkImportCell, with the primary file main.js. main.js provides the Jupyter requirement, load_ipython_extension (see here for details). In this case, that function does two things.

It iterates over the cells in the existing notebook for what would be identified as bulk import cells based on their metadata.
It binds an event that turns newly created cells into bulk import cells based on their metadata.
The Bulk Import Cell itself is a basic class that initializes itself (i.e. sets up the metadata that lets the Narrative know that it's a bulk import cell, only run once per created cell), and sets itself up for running (run each time the page gets loaded). It also handles a few tricky things like setting up icons, and setting up a bus for event listening.
The second part is the Narrative tweaks to create a new bulk import cell. I added a function to the main Narrative module to create one, and bound it to the Import Selected button in the staging area.

None of this is perfect, but it's a place to start.

* Please include a summary of the change and which issue is fixed. 
* Please also include relevant motivation and context.
* List any dependencies that are required for this change.

# Jira Ticket / Issue #
e.g. https://kbase-jira.atlassian.net/browse/DATAUP-X
- [x] Added the Jira Ticket to the title of the PR e.g. (DATAUP-69 Adds a PR template)

# Testing Instructions
* Details for how to test the PR: 
1. install the narrative again (or run the nbextensions/install.sh script) to install the new extension
2. open a local narrative
3. go to the staging area, select a file, and click "Import Selected"

- [x] Tests pass in travis and locally 
- [x] Changes available by spinning up a local narrative and navigating to _X_ to see _Y_

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/truss.works/kbasetruss/development
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: kbase-extension/static/kbase/js/common/error.js [hunk 1]
```diff
@@ -78,9 +80,38 @@ define([
         }
     }
 
+    function reportCellError(title, preamble, error) {
+        const t = html.tag,
+            div = t('div'),
+            p = t('p');
```
Comment (leia-sefkin) [inline kbase-extension/static/kbase/js/common/error.js:86]: small thing but could you add more descriptive names for p and t? single letter variable names make it a bit hard to parse what's going on

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 2]
```diff
@@ -760,75 +760,99 @@ define([
...
+                            bulkMapping[importType] = [];
+                        }
+                        bulkMapping[importType].push(importFile);
+                    }
+                    else {
+                        stagingAreaViewer.initImportApp(importType, importFile);
+                    }
                 });
-          
+            if (bulkMapping) {
```
Comment (leia-sefkin) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:807]: won't this always return true since we have it defined as an obj on line 791?

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js [hunk 3]
```diff
@@ -760,75 +760,99 @@ define([
...
+                Jupyter.narrative.insertBulkImportCell(bulkMapping);
+            }
+            Jupyter.narrative.hideOverlay();
```
Comment (leia-sefkin) [inline kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:810]: not sure if it matters - will this get called twice in the case where we have only individual file imports (and no bulk)?

File: nbextensions/bulkImportCell/bulkImportCell.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+         * present.
+         *
+         * If initialize is falsy, no changes are made to the structure, even if typesToFiles is
+         * present and contains new data.
+         * @param {Cell} cell a Jupyter notebook cell. This should be a code cell, and will throw
+         * an Error if it is not.
+         * @param {boolean} initialize if true, this will initialize the bulk import cell
+         * structure that gets serialized in the cell metadata. This should ONLY be set true when
+         * creating a new bulk import cell, not loading a narrative that already contains one
+         * @param {object} typesToFiles keys = data types, values = arrays of files to import
```
Comment (leia-sefkin) [inline nbextensions/bulkImportCell/bulkImportCell.js:38]: small doc note: would be helpful to have an example here of a data type, and more descriptors on the files (e.g. string which references the file in staging)

File: nbextensions/bulkImportCell/bulkImportCell.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+         */
+        constructor(cell, initialize, typesToFiles) {
+            if (cell.cell_type !== 'code') {
+                throw new Error('Can only create Bulk Import Cells out of code cells!');
+            }
+            this.cell = cell;
+            this.kbaseNode = null;  // this is the
```
Comment (leia-sefkin) [inline nbextensions/bulkImportCell/bulkImportCell.js:45]: suspense in the comments!

File: nbextensions/bulkImportCell/bulkImportCell.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+                return false;
+            }
+            return cell.metadata.kbase.type === CELL_TYPE;
+        }
+
+        /**
+         * Does the initial pass on newly created cells to initialize its metadata and get it
+         * set up for a new life as a Bulk Import Cell.
+         *
+         * @param {object} typesToFiles a map from object type to an array of files.
```
Comment (leia-sefkin) [inline nbextensions/bulkImportCell/bulkImportCell.js:72]: doc note: can you specify what object type you mean here? kbase object type? data type?

File: nbextensions/bulkImportCell/main.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            .then(() => {
+                $([Jupyter.events]).on('insertedAtIndex.Cell', (event, payload) => {
+                    const cell = payload.cell,
+                        setupData = payload.data,
+                        jupyterCellType = payload.type;
+
+                    if (jupyterCellType !== 'code' ||
+                        !setupData ||
+                        !(setupData.type === CELL_TYPE)) {
+                        return;
```
Comment (leia-sefkin) [inline nbextensions/bulkImportCell/main.js:73]: should this return an error?

**Review by leia-sefkin — APPROVED** [review]
some minor points of feedback but it's looking good overall. nice framework to get us started!

File: kbase-extension/static/kbase/js/common/error.js (hunk 1 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/common/error.js:86]: I can change `t` (which I guess should be `tag`), but that's supposed to be a generator for DOM elements. That whole library is effectively a wrapper around HTML. It's used further down like:
```
p('some text')
```
which would generate
```
<p>some text</p>
```

I would normally agree that single-letter names are harder to read than not, but these are intended to be standins for DOM elements.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js (hunk 2 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:807]: Oops, yeah, probably. I meant to have something more descriptive there.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js (hunk 3 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js:810]: Nope, it shouldn't. The flow is supposed to be:
1. make the `bulkMapping` empty object.
2. iterate over the files, add them to the object if they're of the right type.
3. if there's objects in there (which'll be easier to tell when the `if` statement is less broken), then make the cell.

Individual imports should still generate one cell each.

File: nbextensions/bulkImportCell/main.js (hunk 7 above)
Comment (briehl) [reply nbextensions/bulkImportCell/main.js:73]: Nope. That's the nature of the pattern we're using in these extensions. Each cell extension is independent and cycles over the notebook cells at startup. If it finds a cell it's not supposed to turn into, say, a Bulk Import Cell, or an App Cell, it just carries on.

Comment (briehl) [reply nbextensions/bulkImportCell/main.js:73]: As we accumulate more of these, it might make sense to have one heavier controller extension that acts as an overall cell specializer, so they only get iterated over once. But it's a quick process and hasn't been a problem yet. And that's a little out of scope for this PR.

File: nbextensions/bulkImportCell/bulkImportCell.js (hunk 5 above)
Comment (briehl) [reply nbextensions/bulkImportCell/bulkImportCell.js]: suspense removed!

File: nbextensions/bulkImportCell/bulkImportCell.js (hunk 6 above)
Comment (briehl) [reply nbextensions/bulkImportCell/bulkImportCell.js]: I updated that with an example.

File: nbextensions/bulkImportCell/bulkImportCell.js (hunk 4 above)
Comment (briehl) [reply nbextensions/bulkImportCell/bulkImportCell.js]: Updated with an example.
Also, this is just a straw-person. I figure we'll need a way to bring in the object types and files that we're having the bulk importer work on, and this seemed like a first step. I'm sure there's a better way to do it.

File: kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js (hunk 2 above)
Comment (briehl) [reply kbase-extension/static/kbase/js/widgets/narrative_core/upload/stagingAreaViewer.js]: Updated with a fix.

File: kbase-extension/static/kbase/js/common/error.js [hunk 8]
```diff
@@ -78,9 +80,50 @@ define([
...
+        const ui = UI.make({
+            node: document.body
+        });
+        ui.showInfoDialog({
+            title: 'Error',
+            body: div({
+                style: {
+                    margin: '10px'
+                }
+            }, [
```
Comment (ialarmedalien) [inline kbase-extension/static/kbase/js/common/error.js:109]: let's give this div a class to avoid inline style declarations

Comment (briehl) [reply kbase-extension/static/kbase/js/common/error.js]: Ok, done. Have a look.

File: kbase-extension/scss/partials/_errorDisplay.scss [hunk 9]
```diff
@@ -29,3 +29,7 @@
   font-size: 125%;
   margin-top: 2rem;
 }
+
+.error-dialog__body {
```
Comment (ialarmedalien) [inline kbase-extension/scss/partials/_errorDisplay.scss:33]: :tada:

**Review by ialarmedalien — APPROVED** [review]
Ship it!

**Comment by briehl** [comment]
> Ship it!

Thanks. Got a couple conflicts to deal with and tests are oddly flaky when running the whole set, then I'll merge.
